internal/gatewayapi/tls.go

```diff
@@ -11,51 +11,100 @@ import (
 	"fmt"
 
 	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
 )
 
 // validateTLSSecretData ensures the cert and key provided in a secret
 // is not malformed and can be properly parsed
-func validateTLSSecretData(secret *corev1.Secret) error {
-	certData := secret.Data[corev1.TLSCertKey]
+func validateTLSSecretsData(secrets []*corev1.Secret, host *v1beta1.Hostname) error {
+	var publicKeyAlgorithm string
+	var parseErr error
 
-	certBlock, _ := pem.Decode(certData)
-	if certBlock == nil {
-		return fmt.Errorf("unable to decode pem data in %s", corev1.TLSCertKey)
-	}
+	pkaSecretSet := make(map[string]string)
+	for _, secret := range secrets {
+		certData := secret.Data[corev1.TLSCertKey]
 
-	_, err := x509.ParseCertificate(certBlock.Bytes)
-	if err != nil {
-		return fmt.Errorf("unable to parse certificate in %s: %w", corev1.TLSCertKey, err)
-	}
+		certBlock, _ := pem.Decode(certData)
+		if certBlock == nil {
+			return fmt.Errorf("unable to decode pem data in %s", corev1.TLSCertKey)
+		}
 
-	keyData := secret.Data[corev1.TLSPrivateKeyKey]
+		cert, err := x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("unable to parse certificate in %s: %w", corev1.TLSCertKey, err)
+		}
+		publicKeyAlgorithm = cert.PublicKeyAlgorithm.String()
 
-	keyBlock, _ := pem.Decode(keyData)
-	if keyBlock == nil {
-		return fmt.Errorf("unable to decode pem data in %s", corev1.TLSPrivateKeyKey)
-	}
+		keyData := secret.Data[corev1.TLSPrivateKeyKey]
 
-	var parseErr error
+		keyBlock, _ := pem.Decode(keyData)
+		if keyBlock == nil {
+			return fmt.Errorf("unable to decode pem data in %s", corev1.TLSPrivateKeyKey)
+		}
 
-	switch keyBlock.Type {
-	case "PRIVATE KEY":
-		_, err := x509.ParsePKCS8PrivateKey(keyBlock.Bytes)
+		matchedFQDN, err := verifyHostname(cert, host)
 		if err != nil {
-			parseErr = fmt.Errorf("unable to parse PKCS8 formatted private key in %s", corev1.TLSPrivateKeyKey)
+			return fmt.Errorf("hostname %s does not match Common Name or DNS Names in the certificate %s", string(*host), corev1.TLSCertKey)
 		}
-	case "RSA PRIVATE KEY":
-		_, err := x509.ParsePKCS1PrivateKey(keyBlock.Bytes)
-		if err != nil {
-			parseErr = fmt.Errorf("unable to parse PKCS1 formatted private key in %s", corev1.TLSPrivateKeyKey)
+		pkaSecretKey := fmt.Sprintf("%s/%s", publicKeyAlgorithm, matchedFQDN)
+
+		// Check whether the public key algorithm and matched certificate FQDN in the referenced secrets are unique.
+		if certFQDN, ok := pkaSecretSet[pkaSecretKey]; ok {
+			return fmt.Errorf("secret %s/%s public key algorithm must be unique. Cerificate FQDN %s has a conficting algorithm [%s]",
+				secret.Namespace, secret.Name, certFQDN, publicKeyAlgorithm)
+
 		}
-	case "EC PRIVATE KEY":
-		_, err := x509.ParseECPrivateKey(keyBlock.Bytes)
-		if err != nil {
-			parseErr = fmt.Errorf("unable to parse EC formatted private key in %s", corev1.TLSPrivateKeyKey)
+		pkaSecretSet[pkaSecretKey] = getCertFQDN(cert)
```

Comment:
can we do without `getCertFQDN` ?
and instead use
```
pkaSecretSet[pkaSecretKey] = matchedFQDN
```

to reduce code, improve coverage ?

Reply:
Hmm, wanted to keep it for logging purposes, but we can simplify it and use only matchedFQDN